<a href="https://colab.research.google.com/github/RomaViraj/APT/blob/master/Applied_Tech_Project_140_Solution_copy_v0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Instructions

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concepts covered in the lesson: 

**Climate Change - Data Visualisation**

---

#### Problem Statement

On this dataset, you have rich data about crime incidents on São Paulo, Brazil which specifies what has happened in the victim's view, as well as a field indicating what objects were lost. 

As a Data Scientist, you have to clean the data properly to animate the visualize the crimes happening in different areas using Choropleth maps which will show crimes on different dates. So, you have to properly analyse each column and perform different operations.

---

### Dataset Description

The DataFrame consists of the following columns:

|Field|Description|
| --- |:---|
| id | Unique ID for each occurence |
| bairro | São Paulo neighborhood |
| created_at | Date when the report was created |
| descricao | Description of what occurred |
| endereco | Address of the ocurrence |
| id.1 | Duplicate Unique ID for each occurence |
| ip_address_origin | IP address of who created the incident |
| latitude | Address latitude |
| longitude | Address longitude |
| registrou_bo | Flag of whatever the victim has formally registered the incident |
| sexo | Sex of the victim |
| tipo_assalto_id | Kind of occurence |
| titulo | Occurrence title |
| uri | Incident URI |
| user_id | Identificator for the user |
| valor_prejuizo | How much was lost in BRL |
| Bicicleta | Was the bike stolen? |
| Bolsa ou Mochila | Was the backpack stolen? |
| Carteira | Was the wallet stolen? |
| Cartão de Crédito | Was the credit card stolen? |
| Celular | Was the phone stolen? |
| Computador | Was the computer stolen? |
| DVD | Was the DVD stolen? |
| Dinheiro | Was the money stolen? |
| Documentos | Was the documents stolen? |
| Equipamento de Som | Was the sound equipment stolen? |
| Estepe | Was the spare car wheel stolen? |
| MP4 ou Ipod | Was the portable music player stolen? |
| Móveis | Was the furniture stolen? |
| Notebook | Was the laptop stolen? |
| Outros | Was the others stolen? |
| Relógio | Was the wristwatch stolen? |
| Som | Was the soundbox stolen? |
| Tablet | Was the tablet stolen? |
| Tv | Was the television stolen? |
| time | Datetime of the incident |

**Dataset source:** https://www.kaggle.com/danlessa/geospatial-sao-paulo-crime-database

**Citation:** ondefuiroubado.com.br, & Danilo Lessa Bernardineli. (2019). Geospatial Sao Paulo Crime Database.

**Dataset link:** https://s3-whjr-curriculum-uploads.whjr.online/281cc4f3-0f0d-4eb9-8124-178cd51186f1.csv

---

### List of Activities
 
**Activity 1:** Import Modules and Read Data

**Activity 2:** Data Cleaning

**Activity 3:** Visualizing the Crime Rates and Money Lost

---

#### Activity 1: Import Modules and Read Data

1. Import the necessary Python modules.

2. Read the data from the dummy variable to create a Pandas DataFrame and go through the necessary data-cleaning process (if required).

In [ ]:
# Import the required modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

# Read the dataset and print first five records.
url = "https://s3-whjr-curriculum-uploads.whjr.online/281cc4f3-0f0d-4eb9-8124-178cd51186f1.csv"
df = pd.read_csv(url)

# Create a DataFrame with only required features.
df.head()

id             bairro  ...    Tv                     time
0  49787            Butantã  ...  True  2010-12-20 20:30:00.000
1  10679           Itaquera  ...   NaN  2012-01-11 14:00:00.000
2  10682           Itaquera  ...   NaN  2012-01-11 14:00:00.000
3   6166            Morumbi  ...   NaN  2012-01-11 18:20:00.000
4   6521  Alto de Pinheiros  ...   NaN  2012-01-11 08:15:00.000

[5 rows x 36 columns]

Get the information about the dataset.

In [ ]:
# Print the information of the DataFrame using the 'info()' function.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12899 entries, 0 to 12898
Data columns (total 36 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  12899 non-null  int64  
 1   bairro              11196 non-null  object 
 2   created_at          12899 non-null  object 
 3   descricao           12898 non-null  object 
 4   endereco            12899 non-null  object 
 5   id.1                12899 non-null  int64  
 6   ip_address_origin   11705 non-null  object 
 7   latitude            12899 non-null  float64
 8   longitude           12899 non-null  float64
 9   registrou_bo        12375 non-null  object 
 10  sexo                12899 non-null  int64  
 11  tipo_assalto_id     12899 non-null  int64  
 12  titulo              12899 non-null  object 
 13  uri                 12899 non-null  object 
 14  user_id             1166 non-null   float64
 15  valor_prejuizo      12899 non-null  float64
 16  Bici

**Q:** Are there any missing values in the DataFrame?

**A:** Yes.




---



#### Activity 2: Data Cleaning

1. Make a new DataFrame with the following columns: 

  `['id', 'bairro', 'latitude', 'longitude', 'valor_prejuizo', 'Bicicleta', 'Bolsa ou Mochila', 'Carteira', 'Cartão de Crédito',  'Celular', 'Computador', 'DVD', 'Dinheiro', 'Documentos', 'Equipamento de Som', 'Estepe', 'MP4 ou Ipod', 'Móveis', 'Notebook', 'Outros', 'Relógio', 'Som', 'Tablet', 'Tv', 'time']`

2. Drop the `NaN` values from the `bairro` column and fill the `NaN` values of other columns with `False`.

3. Convert `True` and `False` into `0` and `1` by multiplying the values with 1. Calculate the total items stolen using the `sum()` function and drop the unnecessary columns.

Make a new DataFrame with the columns.

In [ ]:
# Create a variable with the column names in a list
features = ['id', 'bairro', 'latitude', 'longitude', 'valor_prejuizo',
       'Bicicleta', 'Bolsa ou Mochila', 'Carteira', 'Cartão de Crédito',
       'Celular', 'Computador', 'DVD', 'Dinheiro', 'Documentos',
       'Equipamento de Som', 'Estepe', 'MP4 ou Ipod', 'Móveis', 'Notebook',
       'Outros', 'Relógio', 'Som', 'Tablet', 'Tv', 'time']

# Create a new DataFrame with the columns selected
crime_df = df[features]
crime_df.head()

id             bairro   latitude  ...  Tablet    Tv                     time
0  49787            Butantã -23.580973  ...    True  True  2010-12-20 20:30:00.000
1  10679           Itaquera -23.540105  ...     NaN   NaN  2012-01-11 14:00:00.000
2  10682           Itaquera -23.541765  ...     NaN   NaN  2012-01-11 14:00:00.000
3   6166            Morumbi -23.611693  ...     NaN   NaN  2012-01-11 18:20:00.000
4   6521  Alto de Pinheiros -23.544564  ...     NaN   NaN  2012-01-11 08:15:00.000

[5 rows x 25 columns]

Drop the `NaN` values from `bairro` column using `dropna()` function. Fill `NaN` values of other columns with 0 using `fillna(False, inplace = True)`.


In [ ]:
# Drop the NaN from 'bairro' column.
crime_df.dropna(subset = ['bairro'], inplace = True)

# Fill the NaN values with False
crime_df.fillna(False, inplace = True)

Display the last 5 rows.

In [ ]:
# Display the last 5 rows of the DataFrame.
crime_df.tail()

id               bairro  ...     Tv                     time
12894  75152  Jardim das Camelias  ...  False  2015-11-17 23:00:00.000
12895  16908      Jardim Paulista  ...  False  2012-03-03 18:30:00.000
12896  78360    Vila Joao Ramalho  ...  False  2012-03-03 13:15:00.000
12897   3080           Campo Belo  ...  False  2012-03-02 21:01:00.000
12898   7462            Pinheiros  ...  False  2012-03-02 09:44:00.000

[5 rows x 25 columns]

Convert `True` and `False` into `0` and `1` by multiplying the values with 1 using `iloc[:,5:-1] * 1` function. 

Calculate the total items stolen using `iloc[:,5:-1].sum(axis = 1)` function and drop the unnecessary columns using `drop(columns = crime_df.iloc[:,5:-2].columns, inplace = True)`.

In [ ]:
# Convert True to 1 and False to 0
crime_df.iloc[:,5:-1] = crime_df.iloc[:,5:-1]*1

# Create a new column for total items stolen
crime_df['total_item_stolen'] = crime_df.iloc[:,5:-1].sum(axis = 1)

# Drop the unnecessary columns
crime_df.drop(columns = crime_df.iloc[:,5:-2].columns, inplace = True)

In [ ]:
crime_df.shape

(11196, 7)

**Q:** What is the dimension of the DataFrame after cleaning the data?

**A:** (11196, 7)



---



#### Activity 3: Visualizing the Crime Rates and Money Lost

1. Convert the `time` column into the datetime data type and extract the year.

2. Visualise the BRL (Brazilian Real Lost) and the total number of items stolen with different locations.

Convert the `time` column into datetime data type and extract the year storing it in a new column in a sorted manner

In [ ]:
# Convert the data type of 'time' column to datetime 
crime_df['time'] = pd.to_datetime(crime_df['time'])

# Extract year from 'time' column using 'year' attribute and 'DatetimeIndex' function of pandas. 
# Store the extracted year as a new column 'year' in the new DataFrame.
crime_df['year'] = pd.DatetimeIndex(crime_df['time']).year

# Sort the new DataFrame by year.
crime_df = crime_df.sort_values(by = "year")

Visualise the BRL (Brazilian Real) and the total number of items stolen with different locations by applying `density_mapbox()` function with `hover_data = ['valor_prejuizo', 'total_item_stolen']` to see both BRL and the total number of items stolen on the above `crime_df` DataFrame.

**Use following code to display the map**: 
```
fig = px.density_mapbox(crime_df, lat = 'latitude', lon = 'longitude', 
                        radius = 5, hover_data = ['valor_prejuizo', 'total_item_stolen'],
                        center = dict(lon = -46.6396, lat = -23.5558),
                        zoom = 8, mapbox_style='stamen-terrain',
                        color_continuous_scale = 'reds',
                        animation_frame = 'year', width=1000, height=500,
                        title = 'BRL (Brazilian Real) Currency lost and Item Lost in São Paulo From 2010 to 2018')
```

In [ ]:
# Import the required library
import plotly.express as px

# Create a figure object using parameters from the respective DataFrame
fig = px.density_mapbox(crime_df, lat = 'latitude', lon = 'longitude', 
                        radius = 5, hover_data = ['valor_prejuizo', 'total_item_stolen'],
                        center = dict(lon = -46.6396, lat = -23.5558),
                        zoom = 8, mapbox_style='stamen-terrain',
                        color_continuous_scale = 'reds',
                        animation_frame = 'year', width=1000, height=500,
                        title = 'BRL (Brazilian Real) Currency lost and Item Lost in São Paulo From 2010 to 2018')

fig.show()

**Q:** What are the variables visible when we hover over a data point?

**A:** `year`, `latitude`, `longitude`, `valor_prejuizo` and `total_item_stolen`

**Note:** Variable names can be different which is given by students.



---

